doc: some misc doxygen changes to polish I3C and userspace APIs and structs

dcpleung opened the pull request:
(See individual commits for details.)

dcpleung committed:
kernel: tls: fix doc on arch_tls_stack_setup()

There was a typo there so fix it.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: demand_paging: add doc to enum arch_page_location

This adds doc to enum arch_page_location.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: userspace: fix and clean up doxygen

This fixes and cleans up the doxygen doc for architecture
specific userspace APIs.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: remove deprecate wording on arch_kernel_init()

The wording on deprecating arch_kernel_init() in favor of prep_c()
has never been materialized. Various architectures are using it to
perform initialization. So remove the wording.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

kartben commented on include/zephyr/drivers/i3c.h:
```diff
@@ -56,19 +56,71 @@ extern "C" {
+ *
+ * 0 - No data bytes following the accepted IBI.
+ * 1 - One data byte (MDB, Mandatory Data Byte) follows the accepted IBI.
+ *     Additional data byes may also follows.
```
```suggestion
 *     Additional data bytes may also follows.
```

dcpleung replied:
Fixed.

kartben commented on include/zephyr/drivers/i3c.h:
```diff
@@ -56,19 +56,71 @@ extern "C" {
  *   - 0: No Limitation
  *   - 1: Limitation obtained via GETMXDS CCC.
  */
+
```
for each group of macros defining the various bits of a register, you may want to consider putting them in `@name { ... }` sections

```c
/**
 * @name Bus Characteristic Register (BCR) fields.
 * @{
 */

...

/** @} */
```

dcpleung replied:
Moved BCR and DCR under named section.

kartben commented on include/zephyr/drivers/i3c.h:
```diff
@@ -56,19 +56,71 @@ extern "C" {
  */
+
+/**
+ * @brief Bus Characteristic Register (BCR) - Max Data Speed Limitation bit.
```
If you go for the suggestion above, you may then want to simplify as the following:

```suggestion
 * @brief Max Data Speed Limitation bit.
```

dcpleung replied:
Changed.

kartben commented on include/zephyr/drivers/i3c.h:
```diff
@@ -56,19 +56,71 @@ extern "C" {
+/**
+ * @brief Bus Characteristic Register (BCR) - Offline Capable bit.
+ *
+ * 0 - Will always response to I3C commands.
+ * 1 - Will not always response to I3C commands.
```
```suggestion
 * 0 - Will always respond to I3C commands.
 * 1 - Will not always respond to I3C commands.
```

dcpleung replied:
Fixed.

kartben commented on include/zephyr/sys/arch_interface.h:
```diff
@@ -787,10 +787,12 @@ int arch_buffer_validate(void *addr, size_t size, int write);
  *
- * @param[in] phys Physical address of region to be mapped, aligned to MMU_PAGE_SIZE
- * @param[in] size Size of region to be mapped, aligned to MMU_PAGE_SIZE
+ * @param[in] phys Physical address of region to be mapped,
+ *                 aligned to CONFIG_MMU_PAGE_SIZE
+ * @param[in] size Size of region to be mapped,
+ *                 aligned to CONFIG_MMU_PAGE_SIZE
```
```suggestion
 * @param[in] phys Physical address of region to be mapped,
 *                 aligned to @kconfig{CONFIG_MMU_PAGE_SIZE}
 * @param[in] size Size of region to be mapped,
 *                 aligned to @kconfig{CONFIG_MMU_PAGE_SIZE}
```

dcpleung replied:
Does it generate link to the kconfig doc? My local build doesn't link anything but simply making it with monospace fonts.

kartben replied:
Yes but only when rendered on the main docs website, not in the Doxygen docs itself

https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/zephyr.doxyfile.in#L278

kartben reviewed (changes requested):
This is awesome - thanks for taking the time to go through this.

Only spotted a few minor typos. With regards to using `@name`d sections, I will leave it up to you to decide if you want to do it. I think this would help make the likes of `ccc.h` easier to quickly parse when browsing the docs, but it's not like everything is not already nicely grouped alphabetically.

dcpleung committed:
i3c: some doxygen API doc polishing

This polishes the doxygen doc on I3C API to make it,
hopefully, more usable. Fixed some typos too.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

kartben reviewed (approved):
Very nice, thanks! 👏🏼
